### ISSUE TITLE
Add syntax highlighting capabilities.

### ISSUE THREAD

**Issue by AmokHuginnsson** [arangodb/linenoise-ng#6]
I would like to know your opinion about extending linenoise-ng
with ability to colorize user input on the fly.

Proposed changes to the API:

```C++
typedef int color_t;
typedef void (linenoiseColorizationCallback)(char const* input, color_t* colors, int size);
void linenoiseSetColorizationCallback(linenoiseColorizationCallback* fn);
```

Where:
`input` is current input
`colors` is buffer for requested colors
`size` is size of both user input in characters and `colors` buffer.

How would it work.
If user want to have colorful input she must simply install "colorization handler".
The handler would be invoked by the library after each change to the input done by
the user. After handler returns library uses data from `colors` buffer to colorize
displayed user input.

Should I code such addition to linenoise-ng?

[Edit]
I have implemented this feature, and here is how it looks like:
![demo](https://drive.google.com/uc?export=download&id=0B53g2Y3z7rWNUVl2YkxPd2lVOVk)
[/Edit]

**Comment by ajg** [comment]
FWIW, this sounds like a useful feature. I'm curious about:
1. What the `color_t` type would look like.
2. How the mapping from `input` to `colors` would work with multi-byte UTF-8 sequences.

**Comment by AmokHuginnsson** [comment]
Hello.

1. I was thinking about `typedef char color_t` or `typedef char unsigned` just not to over do this.
API could also contain color(attribute) definitions, something like:
```C++
struct COLORS {
  enum {
    FG_BLACK         = 0,
    FG_RED           = 1,
    FG_GREEN         = 2,
    FG_BROWN         = 3,
    FG_BLUE          = 4,
    FG_MAGENTA       = 5,
    FG_CYAN          = 6,
    FG_LIGHTGRAY     = 7,
    FG_BOLD          = 8,
    FG_GRAY          = FG_BLACK     | FG_BOLD,
    FG_BRIGHTRED     = FG_RED       | FG_BOLD,
    FG_BRIGHTGREEN   = FG_GREEN     | FG_BOLD,
    FG_YELLOW        = FG_BROWN     | FG_BOLD,
    FG_BRIGHTBLUE    = FG_BLUE      | FG_BOLD,
    FG_BRIGHTMAGENTA = FG_MAGENTA   | FG_BOLD,
    FG_BRIGHTCYAN    = FG_CYAN      | FG_BOLD,
    FG_WHITE         = FG_LIGHTGRAY | FG_BOLD,
    FG_MASK          = 0x07,

    BG_BLACK         = 0,
    BG_RED           = 16,
    BG_GREEN         = 32,
    BG_BROWN         = 48,
    BG_BLUE          = 64,
    BG_MAGENTA       = 80,
    BG_CYAN          = 96,
    BG_LIGHTGRAY     = 112,
    BG_BLINK         = 128,
    BG_GRAY          = BG_BLACK     | BG_BLINK,
    BG_BRIGHTRED     = BG_RED       | BG_BLINK,
    BG_BRIGHTGREEN   = BG_GREEN     | BG_BLINK,
    BG_YELLOW        = BG_BROWN     | BG_BLINK,
    BG_BRIGHTBLUE    = BG_BLUE      | BG_BLINK,
    BG_BRIGHTMAGENTA = BG_MAGENTA   | BG_BLINK,
    BG_BRIGHTCYAN    = BG_CYAN      | BG_BLINK,
    BG_WHITE         = BG_LIGHTGRAY | BG_BLINK,
    BG_MASK          = 0x70,
    ATTR_NORMAL      = ( FG_LIGHTGRAY | BG_BLACK ),
    ATTR_DEFAULT     = -1
  };
};
```

2. I assumed linenoise-ng would be concerned only about displayed characters, so
it would be quite possible that user would get buffers (`input` and `color`) of different sizes.
It would be an user task to handle multi-byte UTF-8 sequences correctly.
In other words. There should be exactly one `color_t` item for one displayed character, no matter of how many bytes this character could be consisted of.

Awaiting you opinion.

**Comment by ajg** [comment]
1. It might be worth it to follow `linenoise`'s [color convention](https://github.com/antirez/linenoise#hints) for compatibility purposes; there, a color is an `int` with the following (xterm) codes...
```
red = 31
green = 32
yellow = 33
blue = 34
magenta = 35
cyan = 36
white = 37;
```
...along with an optional bold (`0` or `1`) attribute.

2. I think rather the `size` should actually be the same for both `input` and `colors`; in the case of UTF-8 either all bytes representing a single code point must be assigned the same color, or simpler still, only the color for the first byte can be used. That way there's no need to do potentially expensive transformations to/from UTF-8 form.

**Comment by AmokHuginnsson** [comment]
I have updated my original message with a simple demo of this feature in action.

**Comment by AmokHuginnsson** [comment]
Due to lack of response here, I decided to create my own fork of this library called [replxx](https://github.com/AmokHuginnsson/replxx). In this fork I added all the improvements I wanted to have. I leave the link here if anyone is interested.

Cheers.
